### PR TITLE
Chore/optimize

### PR DESCRIPTION
**Pull request by SimonSchick**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [ ] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/master/CONTRIBUTING.md)?

### Description of change

This PR does a couple of things (please advise if I should split this into separate PRs) that are all somewhat related.

1)

Replace various operations with faster equivalents (contains some minor `else if` simplification as well.
Main component is the replacement of `_.memoize` which was used to cache array lookups with shiny `Set`s. This also removes some dead code.

2)

While debugging and issue I got really frustrated with the amount of conditional nesting is going on
I reduced the amount of `return } else if (...)` by roughly 50% (mostly using return instead).
In `query.js` ~I also wrapped some promises that were doubled wrapped and nested for some reason, this code is drastically simpler now~ This broke stuff, I still moved the post query logic to a method to make it easier to read.

3)

NOTE: I switched the test reporter to `dot` as that is slightly faster and doesn't create so much useless output, please mention if this is undesirable.
Replaces all occurences of `new Buffer` and `sinon.sandbox.create`  and stubbed out (most) data-type related warnings that were spamming logs.
~Also refactored `ABSTRACT` prototype class to class since it's never directly instantiated.~ Nvm see https://github.com/nodejs/node/issues/23960

4)

Previously forgot to port this :)

### CODE REVIEW

File: lib/dialects/mssql/query.js [hunk 1]
```diff
@@ -62,81 +62,87 @@ class Query extends AbstractQuery {
 
     return new Promise((resolve, reject) => {
       // TRANSACTION SUPPORT
-      if (_.startsWith(this.sql, 'BEGIN TRANSACTION')) {
+      if (this.sql.startsWith('BEGIN TRANSACTION')) {
         connection.beginTransaction(err => {
```
Comment (sushantdhiman) [inline lib/dialects/mssql/query.js]: Directly use return here rather than an extra line below each if block

File: lib/dialects/mssql/query.js [hunk 2]
```diff
@@ -307,9 +325,7 @@ class Query extends AbstractQuery {
 
     match = err.message.match(/Failed on step '(.*)'.Could not create constraint. See previous errors./) ||
       err.message.match(/The DELETE statement conflicted with the REFERENCE constraint "(.*)". The conflict occurred in database "(.*)", table "(.*)", column '(.*)'./) ||
-      err.message.match(/The INSERT statement conflicted with the FOREIGN KEY constraint "(.*)". The conflict occurred in database "(.*)", table "(.*)", column '(.*)'./) ||
-      err.message.match(/The MERGE statement conflicted with the FOREIGN KEY constraint "(.*)". The conflict occurred in database "(.*)", table "(.*)", column '(.*)'./) ||
-      err.message.match(/The UPDATE statement conflicted with the FOREIGN KEY constraint "(.*)". The conflict occurred in database "(.*)", table "(.*)", column '(.*)'./);
+      err.message.match(/The (?:INSERT|MERGE|UPDATE) statement conflicted with the FOREIGN KEY constraint "(.*)". The conflict occurred in database "(.*)", table "(.*)", column '(.*)'./);
```
Comment (sushantdhiman) [inline lib/dialects/mssql/query.js]: :+1:

File: lib/dialects/sqlite/query.js [hunk 3]
```diff
@@ -77,6 +77,162 @@ class Query extends AbstractQuery {
     return ret;
   }
 
+  _handleAfterQuery(metaData, columnTypes, benchmark, queryBegin, err, results) {
```
Comment (sushantdhiman) [inline lib/dialects/sqlite/query.js]: `_handleQueryResponse`

File: lib/dialects/sqlite/query.js [hunk 4]
```diff
@@ -104,162 +260,28 @@ class Query extends AbstractQuery {
...
-                    result = [result, metaData.changes];
-                  }
-
-                  resolve(result);
-                }
+          }
+          resolve(new Promise((resolve, reject) => {
+            const query = this;
+            // cannot use arrow function here because the function is bound to the statement
+            function afterExecute(err, results) {
```
Comment (sushantdhiman) [inline lib/dialects/sqlite/query.js]: ```afterExecute(executionError, ```

File: lib/dialects/sqlite/query.js [hunk 5]
```diff
@@ -104,162 +260,28 @@ class Query extends AbstractQuery {
...
+              try {
+                resolve(query._handleAfterQuery(this, columnTypes, benchmark, queryBegin, err, results));
+                return;
+              } catch (err2) {
```
Comment (sushantdhiman) [inline lib/dialects/sqlite/query.js]: `error`

File: lib/model.js [hunk 6]
```diff
@@ -1090,13 +1090,10 @@ class Model {
     this._dataTypeChanges = {};
     this._dataTypeSanitizers = {};
 
-    this._booleanAttributes = [];
```
Comment (sushantdhiman) [inline lib/model.js]: Keep all these stores, no need to use memoization, you can use `Set`

File: lib/model.js [hunk 7]
```diff
@@ -1090,13 +1090,10 @@ class Model {
...
-    this._dateAttributes = [];
-    this._hstoreAttributes = [];
-    this._rangeAttributes = [];
-    this._jsonAttributes = [];
-    this._geometryAttributes = [];
-    this._virtualAttributes = [];
+    this._hasBooleanAttributes = false;
+    this._hasDateOrBoolAttributes = false;
```
Comment (sushantdhiman) [inline lib/model.js]: Keep `_hasDateAttributes` separate

File: lib/model.js [hunk 8]
```diff
@@ -3024,13 +2998,14 @@ class Model {
         return [affectedRows];
       });
     }).tap(result => {
-      if (options.individualHooks) {
```
Comment (sushantdhiman) [inline lib/model.js]: Remove this change, old version looks clean to me

File: lib/model.js (hunk 6 above)
Comment (sushantdhiman) [reply lib/model.js]: Actually get rid of these changes, better to target them in separate PR. Let's focus on code cleanup in this PR

File: lib/utils.js [hunk 9]
```diff
@@ -42,7 +41,8 @@ function underscoredIf(str, condition) {
 exports.underscoredIf = underscoredIf;
 
 function isPrimitive(val) {
-  return primitives.includes(typeof val);
+  const t = typeof val;
```
Comment (sushantdhiman) [inline lib/utils.js]: Call this `type`

File: package.json [hunk 10]
```diff
@@ -107,14 +107,14 @@
     "test-docker-integration": "env-cmd $npm_package_options_env_cmd npm run test-integration",
     "docs": "esdoc && cp docs/ROUTER esdoc/ROUTER",
     "teaser": "node scripts/teaser",
-    "test-unit": "mocha --require scripts/mocha-bootload --globals setImmediate,clearImmediate --ui tdd --exit --check-leaks --colors -t 30000 --reporter spec \"test/unit/**/*.js\"",
+    "test-unit": "mocha --require scripts/mocha-bootload --globals setImmediate,clearImmediate --ui tdd --exit --check-leaks --colors -t 30000 --reporter dot \"test/unit/**/*.js\"",
```
Comment (sushantdhiman) [inline package.json]: No need to change reporter, Please use spec

File: test/unit/sql/data-types.test.js [hunk 11]
```diff
@@ -9,11 +9,22 @@ const Support = require('../support'),
...
-  dialect = Support.getTestDialect();
+  dialect = Support.getTestDialect(),
+  logger = require('../../../lib/utils/logger'),
+  { stub } = require('sinon');
 
 // Notice: [] will be replaced by dialect specific tick/quote character when there is not dialect specific expectation but only a default expectation
 
 describe(Support.getTestDialectTeaser('SQL'), () => {
+
+  let warnStub;
```
Comment (sushantdhiman) [inline test/unit/sql/data-types.test.js]: Why this change? If just to remove warning logs, please revert it

File: lib/data-types.js [hunk 12]
```diff
@@ -406,24 +406,29 @@ DECIMAL.prototype.validate = function validate(value) {
   return true;
 };
 
-for (const floating of [FLOAT, DOUBLE, REAL]) {
-  floating.prototype.escape = false;
-  floating.prototype._value = function _value(value) {
+const protoExtensions = {
```
[the remark this answers is not in the document]
Comment (SimonSchick) [reply lib/data-types.js]: Avoid re-creating the same functions.

**Comment by SimonSchick** [comment]
Addressed all comments and squashed all commits since there isn't much point in keeping them separate anymore, waiting for CI build to pass now.

**Comment by SimonSchick** [comment]
@sushantdhiman we good now?

File: lib/model.js [hunk 13]
```diff
@@ -3024,13 +3028,14 @@ class Model {
...
-        return Promise.map(instances, instance => {
-          return this.runHooks('afterUpdate', instance, options);
-        }).then(() => {
-          result[1] = instances;
-        });
+      if (!options.individualHooks) {
```
Comment (sushantdhiman) [inline lib/model.js]: Please revert this change old code was simple

File: lib/model.js [hunk 14]
```diff
@@ -992,33 +993,32 @@ class Model {
       .map(index => Utils.nameIndex(index, this.getTableName()));
 
     this.primaryKeys = {};
-    this._readOnlyAttributes = [];
+    this._readOnlyAttributes = new Set();
```
Comment (sushantdhiman) [inline lib/model.js]: Keep this as array for now

Comment (sushantdhiman) [reply lib/model.js]: So we match other internal data stores

File: lib/utils.js [hunk 15]
```diff
@@ -644,4 +640,16 @@ function nameIndex(index, tableName) {
 }
 exports.nameIndex = nameIndex;
 
-
+/**
```
Comment (sushantdhiman) [inline lib/utils.js]: Unused, please remove

**Comment by SimonSchick** [comment]
done

**Comment by sushantdhiman** [comment]
Thanks @SimonSchick
